### PR TITLE
Composable clocks

### PR DESCRIPTION
**Pull request by thenewwazoo**
Getting a PR up early for feedback. Things I'd like to do in this PR, but that we may be able to defer:
* migrate to the new `power` module and pass PWR registers around
* add CR to RCC and add helper functions
* pass BDCR into `configure` functions instead of reaching into the register directly

I haven't extensively tested this branch, but it seems to work.

### CODE REVIEW

File: src/lcd/mod.rs [hunk 1]
```diff
@@ -120,11 +120,9 @@ impl LCD {
...
     ///HSE clock is not supported yet...
     pub fn validate(lcd: &mut stm32l4x6::LCD, bdcr: &mut BDCR, configuration: &config::Config) -> ValidationResult {
-        let clock_frequency: u32 = match bdcr.rtc_clock() {
-            RtcClockType::None => return ValidationResult::ClockNotSet,
-            RtcClockType::LSE => 32_000,
-            RtcClockType::LSI => 32_768,
-            RtcClockType::HSE => panic!("HSE clock is not supported yet")
+        let clock_frequency: u32 = match bdcr.rtc_clock().freq(None) {
+            Ok(f) => f,
+            Err(_) => { return ValidationResult::ClockNotSet; },
```
Comment (DoumanAsh) [inline src/lcd/mod.rs]: { } to wrap return is not necessary alongside with ; afaik

File: src/rcc.rs [hunk 2]
```diff
@@ -1,26 +1,341 @@
...
+use stm32l4x6::{rcc, RCC, PWR};
 
 use common::Constrain;
 use time::Hertz;
 use flash::ACR;
 
 impl Constrain<Rcc> for RCC {
+    /// Create an RCC peripheral handle.
+    ///
+    /// Per 6.2, "The MSI is used as system clock source after startup from Reset, configured
```
Comment (DoumanAsh) [inline src/rcc.rs]: Is it the same as with turning power off/on?

Comment (DoumanAsh) [inline src/rcc.rs]: Please instead write along following line.

According to Reference Manual Ch. 6.2:
The default System Clock source is MSI clock with frequency 4 MHz

File: src/rcc.rs [hunk 3]
```diff
@@ -1,26 +1,341 @@
...
+    /// at 4 MHz."
     fn constrain(self) -> Rcc {
+        // Enable write access to the BDCR
+        unsafe { (*PWR::ptr()).cr1.modify(|_,w| w.dbp().set_bit()); }
```
Comment (DoumanAsh) [inline src/rcc.rs]: afaik isn't BDCR is not necessary to be written.
Or do you think it is ok to lift write protection even if we are not going to write there?

Please add the comment about it to documentation on RCC struct

File: src/rcc.rs [hunk 4]
```diff
@@ -1,26 +1,341 @@
...
+                pclk2: None,
+                sysclk: clocking::SysClkSource::MSI(
+                    clocking::MediumSpeedInternalRC::new(4_000_000, false)
+                    ),
+            }
         }
     }
 }
 
+pub mod clocking {
```
Comment (DoumanAsh) [inline src/rcc.rs]: I suggest to split rcc.rs into several files.
The structure can be something like
`src/rcc/mod.rs`
`src/rcc/clocking.rs`

File: src/rcc.rs [hunk 5]
```diff
@@ -1,26 +1,341 @@
...
+
+    use super::rcc;
+
+    pub trait InputClock {
+        fn freq(&self) -> u32;
+    }
+
+    /// High-speed internal 16 MHz RC
+    #[derive(Clone, Copy)]
+    pub struct HighSpeedInternal16RC {
```
Comment (DoumanAsh) [inline src/rcc.rs]: I think maybe it would be better to call it just Clock instead of RC?

File: src/rcc.rs [hunk 6]
```diff
@@ -1,26 +1,341 @@
...
+        pub always_on: bool,
+        pub auto_start: bool,
+    }
+
+    impl HighSpeedInternal16RC {
+        pub fn configure(&self, rcc: &rcc::RegisterBlock) -> (u32, u8) {
```
Comment (DoumanAsh) [inline src/rcc.rs]: Add documentation for function

File: src/rcc.rs [hunk 7]
```diff
@@ -1,26 +1,341 @@
...
+                         );
+            while rcc.cr.read().hsirdy().bit_is_clear() {}
+            (16_000_000, 0b01)
+        }
+    }
+
+    /// Medium-speed internal 100 kHz - 48 MHz RC
+    #[derive(Clone, Copy)]
+    pub struct MediumSpeedInternalRC {
+        f: u32,
```
Comment (DoumanAsh) [inline src/rcc.rs]: Please instead of `f` let it be `freq` so that it would be clear :)

File: src/rcc.rs [hunk 8]
```diff
@@ -1,26 +1,341 @@
...
+        auto_cal: bool,
+        bits: u8,
+    }
+
+    impl MediumSpeedInternalRC {
+
+        pub fn new(f: u32, a: bool) -> Self {
+            let bits = match f {
+                100_000 =>    0b0000,
```
Comment (DoumanAsh) [inline src/rcc.rs]: Can you please add reference to manual chapter from where these values come from?

File: src/rcc.rs [hunk 9]
```diff
@@ -1,26 +1,341 @@
...
+            rcc.cr.modify(|_, w| unsafe {
+                w.msirange()
+                    .bits(self.bits())
+                    .msirgsel()
+                    .set_bit()
+            });
+            while rcc.cr.read().msirdy().bit_is_clear() {}
+
+            if self.auto_cal {
+                // FIXME This... may not work? I'm not sure if I've got a board problem or using
```
Comment (DoumanAsh) [inline src/rcc.rs]: Could it be it requires https://github.com/DoumanAsh/stm32l4x6_hal/blob/master/src/power.rs#L45 ?

File: src/rcc.rs [hunk 10]
```diff
@@ -1,26 +1,341 @@
...
+    }
+
+    pub enum RtcClkSource {
+        None,
+        /// Internal 32 kHz RC
+        LSI,
+        /// External 32.768 kHz oscillator
+        LSE,
+        HSEDiv32,
+    }
```
Comment (DoumanAsh) [inline src/rcc.rs]: Add empty line after }

File: src/rcc.rs [hunk 11]
```diff
@@ -1,26 +1,341 @@
...
+        }
+    }
+
+    pub enum RtcClkSource {
+        None,
+        /// Internal 32 kHz RC
+        LSI,
+        /// External 32.768 kHz oscillator
+        LSE,
+        HSEDiv32,
```
Comment (DoumanAsh) [inline src/rcc.rs]: Describe HSE variant?

File: src/rcc.rs [hunk 12]
```diff
@@ -1,26 +1,341 @@
...
+    }
+    impl RtcClkSource {
+        pub fn freq(&self, hse: Option<HighSpeedExternalOSC>) -> Result<u32, ()> {
```
Comment (DoumanAsh) [inline src/rcc.rs]: I think Option<u32> works better here.
Having None is not error, it is just has no value.

Also please don't forget to add documentation for all Public methods

File: src/rcc.rs [hunk 13]
```diff
@@ -1,26 +1,341 @@
...
+                    if let Some(clk) = hse {
+                        Ok(clk.freq() / 32)
+                    } else {
+                        Err(())
+                    }
+                },
+            }
+        }
+
+        pub fn bits(&self) -> u8 {
```
Comment (DoumanAsh) [inline src/rcc.rs]: Maybe use `repr(u8)` as I originally had with C style enum?
This way you can just make `self as u8`?

File: src/rcc.rs [hunk 14]
```diff
@@ -1,26 +1,341 @@
...
+
+    }
+
+    #[derive(Clone, Copy)]
+    pub enum SysClkSource {
+        HSI16(HighSpeedInternal16RC),
+        MSI(MediumSpeedInternalRC),
+        HSE(HighSpeedExternalOSC),
+        PLL(PLLClkOutput),
+    }
```
Comment (DoumanAsh) [inline src/rcc.rs]: Empty line after }

File: src/rcc.rs [hunk 15]
```diff
@@ -1,26 +1,341 @@
...
+    pub struct PLLClkOutput {
+        pub src: PLLClkSource,
+        pub m: u8,
+        n: u8,
+        r: u8,
+        f: u32,
+    }
+
+    impl PLLClkOutput {
+        pub fn new(src: PLLClkSource, m: u8, n: u8, r: u8) -> Self {
```
Comment (DoumanAsh) [inline src/rcc.rs]: Certainly need description and more descriptive names for arguments!

Please add reference to manual's chapter that describes configuration for PLL clock

File: src/rcc.rs [hunk 16]
```diff
@@ -1,26 +1,341 @@
...
+
+            PLLClkOutput {src, m, n, r, f}
+        }
+
+        pub fn configure(&self, rcc: &rcc::RegisterBlock) -> (u32, u8) {
+            let pllsrc_bits = self.src.configure(rcc);
+            rcc.cr.modify(|_,w| w.pllon().clear_bit());
+            while rcc.cr.read().pllrdy().bit_is_set() {}
+            rcc.pllcfgr.modify(|_,w| unsafe {
+                w
```
Comment (DoumanAsh) [inline src/rcc.rs]: I don't really like this strange way to put first method on next line.
Maybe just have 
```
w.<method>()
  .<method2>()
```

File: src/rcc.rs [hunk 17]
```diff
@@ -1,26 +1,341 @@
...
+                    .bits(pllsrc_bits)
+                    .pllm()
+                    .bits(self.m - 1)
+                    .plln()
+                    .bits(self.n)
+                    .pllr()
+                    .bits(self.r)
+            });
+            rcc.cr.modify(|_,w| w.pllon().set_bit());
+            rcc.pllcfgr.modify(|_,w| w.pllren().set_bit());
```
Comment (DoumanAsh) [inline src/rcc.rs]: Don't we need to wait for PLL t get ready?

**Review by DoumanAsh — COMMENTED** [review]
Nice stuff you're doing here 👍

**Comment by thenewwazoo** [comment]
Thanks for the feedback! I'll clear the style notes shortly (probably just run it through `rustfmt` unless you object).

What do you think of the concept itself? I'm adding U(S)ART support right now, and the U(S)ARTs take a clock input, which will be selected similarly to this (`rcc::clocking::USARTClkSource`). I think once I have a solid example of how it's used "in the field" (outside of the narrow domain of RCC), it will help the API mature.

**Comment by DoumanAsh** [comment]
I really like concept, sadly I don't have great deal of experience with stm32 so I cannot say my feedback is 100% good enough 😅

File: src/rcc.rs (hunk 14 above)
Comment (thenewwazoo) [reply src/rcc.rs]: I ran a few files through rustfmt, so the change got a bit noisier

File: src/rcc.rs (hunk 15 above)
Comment (thenewwazoo) [reply src/rcc.rs]: Added a comment - in this case, `m`, `n`, and `r` are really the argument names :D

File: src/rcc.rs (hunk 5 above)
Comment (thenewwazoo) [reply src/rcc.rs]: I tried to make the names match the reference manual as much as possible

**Comment by thenewwazoo** [comment]
I'm going to call this one baked. Fixes #1

**Comment by DoumanAsh** [comment]
I'll review later on, but if we're going to ust rustfmt I'd like to have some configuration.
I don't really like how this tool breaks long lines into some shitty multi-line :(

I'll look into its configuration and will push it through for use

**Comment by DoumanAsh** [comment]
Can you please re-base on this commit https://github.com/DoumanAsh/stm32l4x6_hal/commit/505dfd4a7793f5d7a9011c9a9a248bf83da2de89 and re-run rustfmt?

File: src/gpio/mod.rs [hunk 18]
```diff
@@ -171,91 +170,142 @@ macro_rules! impl_pin {
...
 
                 $PXi(PhantomData)
             }
 
             // alternate function mode
             fn set_alt_fun_mode(moder: &mut MODER<$GPIOX>, afr: &mut $AFR<$GPIOX>, af: u32) {
-                moder.moder().modify(|r, w| unsafe { w.bits((r.bits() & !(0b11 << Self::OFFSET)) | (0b10 << Self::OFFSET)) });
-                afr.afr().modify(|r, w| unsafe { w.bits((r.bits() & !(0b1111 << Self::OFFSET)) | (af << Self::OFFSET)) });
+                moder.moder().modify(|r, w| unsafe {
+                    w.bits((r.bits() & !(0b11 << Self::OFFSET)) | (0b10 << Self::OFFSET))
```
Comment (DoumanAsh) [inline src/gpio/mod.rs:248]: I remember you mentioned error in alternative function mode, will you make a separate PR for that?

**UPD**: Ignore it, I saw the change in your next PR

File: src/rcc/clocking.rs [hunk 19]
```diff
@@ -0,0 +1,292 @@
...
+use super::rcc;
+
+pub trait InputClock {
+    fn freq(&self) -> u32;
+}
+
+/// High-speed internal 16 MHz RC
+#[derive(Clone, Copy)]
+pub struct HighSpeedInternal16RC {
+    pub always_on: bool,
```
Comment (DoumanAsh) [inline src/rcc/clocking.rs]: Please describe public fields here and for all other structs in this module

File: src/rcc/clocking.rs [hunk 20]
```diff
@@ -0,0 +1,292 @@
...
+}
+
+impl MediumSpeedInternalRC {
+    pub fn new(freq: u32, auto_cal: bool) -> Self {
+        MediumSpeedInternalRC { freq, auto_cal }
+    }
+
+    pub fn bits(&self) -> u8 {
+        match self.freq {
+            100_000 => 0b0000,
```
Comment (DoumanAsh) [inline src/rcc/clocking.rs]: Can you please add reference to manual's chapter? :)

File: src/rcc/clocking.rs [hunk 21]
```diff
@@ -0,0 +1,292 @@
...
+/// Medium-speed internal 100 kHz - 48 MHz RC
+#[derive(Clone, Copy)]
+pub struct MediumSpeedInternalRC {
+    freq: u32,
+    auto_cal: bool,
+}
+
+impl MediumSpeedInternalRC {
+    pub fn new(freq: u32, auto_cal: bool) -> Self {
+        MediumSpeedInternalRC { freq, auto_cal }
```
Comment (DoumanAsh) [inline src/rcc/clocking.rs]: Should we maybe validate frequency right here?

File: src/rcc/clocking.rs [hunk 22]
```diff
@@ -0,0 +1,292 @@
...
+
+    /// Configures the MSI to the specified frequency, and enables hardware
+    /// auto-calibration if requested by enabling (and waiting for) the LSE.
+    pub fn configure(&self, rcc: &rcc::RegisterBlock) -> (u32, u8) {
+        rcc.cr
+            .modify(|_, w| unsafe { w.msirange().bits(self.bits()).msirgsel().set_bit() });
+        while rcc.cr.read().msirdy().bit_is_clear() {}
+
+        if self.auto_cal {
+            // FIXME This... may not work? I'm not sure if I've got a board problem or using
```
Comment (DoumanAsh) [inline src/rcc/clocking.rs:103]: Is this still an issue?

If LSE is not ready then mot likely you failed to enable it.
You can try to check register's value after you enable it.
The only reason I can think of, is that BDCR write protection is not lifted...

File: src/rcc/clocking.rs [hunk 23]
```diff
@@ -0,0 +1,292 @@
...
+impl InputClock for HighSpeedExternalOSC {
+    fn freq(&self) -> u32 {
+        self.0
+    }
+}
+
+impl HighSpeedExternalOSC {
+    /// Turns on the HSE oscillator.
+    ///
+    /// (Should this also configure the pin?)
```
Comment (DoumanAsh) [inline src/rcc/clocking.rs:138]: I cannot remember right now which GPIO pins are required for HSE, but if you unsure right, you can maybe leave comments here like: "Following GPIO pins must configured as <describe how>: <enumeratino of pins>"

But generally unless it is too complex, I think we should do this in user stead

File: src/rcc/clocking.rs [hunk 24]
```diff
@@ -0,0 +1,292 @@
...
+    None,
+    /// Internal 32 kHz RC
+    LSI,
+    /// External 32.768 kHz oscillator
+    LSE,
+    /// High-speed external oscillator, prescaled by (a fixed value of) 32
+    HSEDiv32,
+}
+
+    impl RtcClkSource {
```
Comment (DoumanAsh) [inline src/rcc/clocking.rs]: Extra spaces

File: src/rcc/clocking.rs [hunk 25]
```diff
@@ -0,0 +1,292 @@
...
+    pub m: u8,
+    n: u8,
+    r: u8,
+    f: u32,
+}
+
+impl PLLClkOutput {
+    /// Create a new PLL clock source to use as an input. The arguments refer to the scale
+    /// factors described in Figs. 15 and 16 of the reference manual, and end up in the PLLM,
+    /// PLLN, and PLLR fields of the PLLCFGR register.
```
Comment (DoumanAsh) [inline src/rcc/clocking.rs]: Please add note that function will panic if resulting frequency exceeds SYS_CLOCK_MAX

File: src/rcc/clocking.rs [hunk 26]
```diff
@@ -0,0 +1,292 @@
...
+
+use super::rcc;
+
+pub trait InputClock {
+    fn freq(&self) -> u32;
+}
+
+/// High-speed internal 16 MHz RC
+#[derive(Clone, Copy)]
+pub struct HighSpeedInternal16RC {
```
Comment (DoumanAsh) [inline src/rcc/clocking.rs:37]: Looking at all clocks I'm wondering: should we maybe make common trait that describes all common methods of configuration?
Such as `configure` and `bits`?

Instead of implementing each struct on its own, we could make them generic over trait like:

```rust
pub trait ClockConfig {
    pub fn configure(&self, rcc: &rcc::RegisterBlock) -> (u32, u8);
    pub fn bits() -> u8;
}
```

or maybe we could make it part of singular Clock trait?

File: src/rcc/clocking.rs [hunk 27]
```diff
@@ -0,0 +1,292 @@
...
+    ///
+    /// (Should this also configure the pin?)
+    pub fn configure(&self, rcc: &rcc::RegisterBlock) -> (u32, u8) {
+        rcc.cr.modify(|_, w| w.hseon().set_bit());
+        while rcc.cr.read().hserdy().bit_is_clear() {}
+        (self.freq(), 0b10)
+    }
+}
+
+pub enum RtcClkSource {
```
Comment (DoumanAsh) [inline src/rcc/clocking.rs]: Why not make this enum `#repr(u8)` and embed bits into its variants?

```rust
pub enum RtcClkSource {
    None = 0,
    LSI = 1,
    LSE = 2,
    HSEDiv32 = 3,
}
```

Then in method `bits` you can just cast it to u8

File: src/rcc/mod.rs [hunk 28]
```diff
@@ -1,22 +1,46 @@
...
+    /// Create an RCC peripheral handle.
+    ///
+    /// Per Reference Manual Ch. 6.2 the default System Clock source is MSI clock with frequency 4 MHz
+    ///
+    /// The `constrain` method enables write access to the BDCR, and the `freeze` method disables
+    /// it again. This is to enable changing LSE- and RTC-related settings.
     fn constrain(self) -> Rcc {
+        // Enable write access to the BDCR; this is necessary to enable the LSE and change RTC
+        // settings.
+        unsafe {
```
Comment (DoumanAsh) [inline src/rcc/mod.rs:27]: There is https://github.com/DoumanAsh/stm32l4x6_hal/blob/master/src/power.rs#L45 for this purpose

Also note afaik you  need to wait for BDCR lock to be lifted(it should be few ticks, but it is still better to wait)

File: src/rcc/mod.rs [hunk 29]
```diff
@@ -317,45 +327,35 @@ impl CFGR {
...
+                .ppre1()
+                .bits(ppre1_bits)
+                .hpre()
+                .bits(hpre_bits)
+                .sw()
+                .bits(sw_bits)
+        });
+
+        // Disable BDCR write access
+        unsafe {
```
Comment (DoumanAsh) [inline src/rcc/mod.rs:349]: It should already done in `contraint` method though?

**Review by DoumanAsh — COMMENTED** [review]
It look nice, I appreciate your hep with tidying up clocks.

File: src/rcc/clocking.rs (hunk 21 above)
Comment (thenewwazoo) [reply src/rcc/clocking.rs]: I made a TODO note here - it's a small change but I'll have to get to it later

File: src/rcc/clocking.rs (hunk 22 above)
Comment (thenewwazoo) [reply src/rcc/clocking.rs:103]: I haven't re-tested this, but will need to return to it soon, so I'll have to tackle it then.

File: src/rcc/mod.rs (hunk 29 above)
Comment (thenewwazoo) [reply src/rcc/mod.rs:349]: Other way - `constrain()` enables write access, `freeze()` disables it
